### PR TITLE
add(root): new package zmap

### LINKED ISSUE #11700: [Package]: zmap
### Package description

ZMap is a fast single packet network scanner designed for Internet-wide network surveys. On a typical desktop computer with a gigabit Ethernet connection, ZMap is capable scanning the entire public IPv4 address space in under 45 minutes. With a 10gigE connection and [PF_RING](http://www.ntop.org/products/packet-capture/pf_ring/), ZMap can scan the IPv4 address space in under 5 minutes.

ZMap operates on GNU/Linux, Mac OS, and BSD. ZMap currently has fully implemented probe modules for TCP SYN scans, ICMP, DNS queries, UPnP, BACNET, and can send a large number of [UDP probes](https://github.com/zmap/zmap/blob/master/examples/udp-probes/README). If you are looking to do more involved scans, e.g., banner grab or TLS handshake, take a look at [ZGrab 2](https://github.com/zmap/zgrab2), ZMap's sister project that performs stateful application-layer handshakes.

https://pkgs.org/download/zmap

### Home page URL

https://github.com/zmap/zmap

### Source code URL

https://github.com/zmap/zmap

### Packaging policy acknowledgement

- [X] I have read and understand the [Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy).

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by nexplorer-3e**
Close #11700 .

### CODE REVIEW

**Comment by licy183** [comment]
Hello, thanks for your contribution! Please rewrite the commit message using format like `add(root): new package zmap` or `new package: zmap`, and when updating branch, do not use `update branch with merge request`, rather than use `update with rebase`.

**Comment by 2096779623** [comment]
Why add `-j`?
If you want to make the build faster, you can add `TERMUX_MAKE_PROCESSES=number of threads` to `build.sh` .
example: https://github.com/termux/termux-packages/blob/master/packages/at/build.sh#L13

**Comment by licy183** [comment]
Thanks! Please squash these commits into one commit. Anyway, have you tested it on device? Is it working fine?

**Comment by nexplorer-3e** [comment]
Working on my phone with interface rmnet. Should work on other interfaces.

________________________________
寄件者: Uchiha Kakashi ***@***.***>
寄件日期: 2022年10月9日 星期日 下午1:28
收件者: termux/termux-packages ***@***.***>
副本: eval Nya ***@***.***>; Author ***@***.***>
主旨: Re: [termux/termux-packages] add(root): new package zmap (PR #12303)


Thanks! Please squash these commits into one commit. Anyway, have you tested it on device? Is it working fine?

File: root-packages/zmap/build.sh [hunk 1]
```diff
@@ -0,0 +1,12 @@
+TERMUX_PKG_HOMEPAGE=https://zmap.io/
+TERMUX_PKG_DESCRIPTION="a fast single packet network scanner designed for Internet-wide network surveys"
+TERMUX_PKG_LICENSE="Apache-2.0"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION="3.0.0-beta1"
+TERMUX_PKG_SRCURL=https://github.com/zmap/zmap/archive/refs/tags/v${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=c97c452930ad6af95ad9ad79e66fea742f904b8e102c185f1001822cc5b5ceb3
+TERMUX_PKG_DEPENDS="libgmp, json-c, libpcap"
+TERMUX_PKG_BUILD_DEPENDS="libgmp, gengetopt, libpcap, flex, byacc, json-c, pkg-config, libunistring"
```
Comment (Grimler91) [inline root-packages/zmap/build.sh:9]: These build dependencies look weird, libraries are normally runtime dependencies, and tools like flex, byacc, json-c etc should be installed through [setup-ubuntu.sh](https://github.com/termux/termux-packages/blob/master/scripts/setup-ubuntu.sh) / [setup-termux.sh](https://github.com/termux/termux-packages/blob/master/scripts/setup-termux.sh)

Probably libraries should be moved to `TERMUX_PKG_DEPENDS`

File: root-packages/zmap/build.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+TERMUX_PKG_BUILD_IN_SRC=true
```
Comment (Grimler91) [inline root-packages/zmap/build.sh:10]: Packages that use CMake most often build fine out-of-src, have you tried without this line?

**Review by Grimler91 — CHANGES_REQUESTED** [review]
Thanks for the PR, see comments for ideas for improvements

File: root-packages/zmap/build.sh (hunk 1 above)
Comment (nexplorer-3e) [reply root-packages/zmap/build.sh]: Looks like `json-c` didn't get installed on preparation. I would added it back later.

**Comment by 2096779623** [comment]
```bash
/data/data/com.termux/files/usr/sbin/zmap
```

Looks like you need to change `sbin` to `bin` .

**Comment by nexplorer-3e** [comment]
> ```shell
> /data/data/com.termux/files/usr/sbin/zmap
> ```
> 
> Looks like you need to change `sbin` to `bin` .

`su` binary in $PREFIX/bin would override the environment variable to `$PATH`, which move binary to `$PREFIX` would make no sense.

**Comment by Grimler91** [comment]
> Looks like you need to change `sbin` to `bin` .

[termux_step_massage moves sbin/* to bin/](https://github.com/termux/termux-packages/blob/master/scripts/build/termux_step_massage.sh#L15-L16) so shouldn't be necessary in this case

**Comment by 2096779623** [comment]
Looks like the hardcoded path in the config file needs to be modified as well.

```bash
.../etc/zmap # cat zmap.conf
### Probe Module to use
#probe-module tcp_synscan


### Destination port to scan
#target-port 443

### Scan rate in packets/sec
#rate 10000

### Scan rate in bandwidth (bits/sec); overrides `rate`
#bandwidth 1M   # 1mbps


### Blocklist file to use. We encourage you to exclude
### RFC1918, IANA reserved, and multicast networks,
### in addition to those who have opted out of your
### network scans.
blocklist-file "/etc/zmap/blocklist.conf"

### Optionally print a summary at the end
#summary
```

**Comment by 2096779623** [comment]
Currently it seems to be working fine on my device.(fixed depends and conf file)

```bash
.../etc/zmap # zmap -p 8022 -i wlan0 127.0.0.1
Oct 10 22:47:56.551 [INFO] zmap: By default, ZMap will output the unique IP addresses of hosts that respond successfully (e.g., SYN-ACK packet). This is equivalent to running ZMap with the following flags: --output-module=csv --output-fields=saddr --output-filter='success=1 && repeat=0' --no-header-row. If you want all responses, explicitly set an output module or set --output-filter="".
 0:00 0%; send: 0 0 p/s (0 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:01 13%; send: 1 done (116 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:02 25%; send: 1 done (116 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:03 38%; send: 1 done (116 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:04 50%; send: 1 done (116 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:05 63% (3s left); send: 1 done (116 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:06 75% (2s left); send: 1 done (116 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:07 88% (1s left); send: 1 done (116 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:08 100% (0s left); send: 1 done (116 p/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
Oct 10 22:48:05.589 [INFO] zmap: completed
```

**Comment by nexplorer-3e** [comment]
Tried just add `\$PREFIX` with no luck. though it just need to work on termux.
patch might come later.

________________________________
寄件者: Kevin Williams ***@***.***>
寄件日期: 2022年10月10日 星期一 下午10:47
收件者: termux/termux-packages ***@***.***>
副本: eval Nya ***@***.***>; Author ***@***.***>
主旨: Re: [termux/termux-packages] add(root): new package zmap (PR #12303)


Looks like the hardcoded path in the config file needs to be modified as well.

.../etc/zmap # cat zmap.conf
### Probe Module to use
#probe-module tcp_synscan


### Destination port to scan
#target-port 443

### Scan rate in packets/sec
#rate 10000

### Scan rate in bandwidth (bits/sec); overrides `rate`
#bandwidth 1M   # 1mbps


### Blocklist file to use. We encourage you to exclude
### RFC1918, IANA reserved, and multicast networks,
### in addition to those who have opted out of your
### network scans.
blocklist-file "/etc/zmap/blocklist.conf"

### Optionally print a summary at the end
#summary

**Comment by 2096779623** [comment]
> Tried just add `\$PREFIX` with no luck. though it just need to work on termux. patch might come later.

Use `@TERMUX_PREFIX@` instead of hardcoded `$PREFIX`.

**Review by 2096779623 — APPROVED** [review]

**Comment by Grimler91** [comment]
Thanks!

**Pull request merged**